File: services/identity-federation-service/src/oauth_integration/models/sso.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct SsoProvider {
    pub id: Uuid,
    pub slug: String,
    pub name: String,
    pub provider_type: String,
    pub enabled: bool,
    pub client_id: Option<String>,
    pub client_secret: Option<String>,
    pub issuer_url: Option<String>,
    pub authorization_url: Option<String>,
    pub token_url: Option<String>,
    pub userinfo_url: Option<String>,
    pub scopes: Vec<String>,
    pub saml_metadata_url: Option<String>,
    pub saml_entity_id: Option<String>,
    pub saml_sso_url: Option<String>,
    pub saml_certificate: Option<String>,
    pub attribute_mapping: Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SsoProviderResponse {
    pub id: Uuid,
    pub slug: String,
    pub name: String,
    pub provider_type: String,
    pub enabled: bool,
    pub client_id: Option<String>,
    pub client_secret_configured: bool,
    pub issuer_url: Option<String>,
    pub authorization_url: Option<String>,
    pub token_url: Option<String>,
    pub userinfo_url: Option<String>,
    pub scopes: Vec<String>,
    pub saml_metadata_url: Option<String>,
    pub saml_entity_id: Option<String>,
    pub saml_sso_url: Option<String>,
    pub attribute_mapping: Value,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl SsoProvider {
    pub fn into_response(self) -> SsoProviderResponse {
        SsoProviderResponse {
            id: self.id,
            slug: self.slug,
            name: self.name,
            provider_type: self.provider_type,
            enabled: self.enabled,
            client_id: self.client_id,
            client_secret_configured: self.client_secret.is_some(),
            issuer_url: self.issuer_url,
            authorization_url: self.authorization_url,
            token_url: self.token_url,
            userinfo_url: self.userinfo_url,
            scopes: self.scopes,
            saml_metadata_url: self.saml_metadata_url,
            saml_entity_id: self.saml_entity_id,
            saml_sso_url: self.saml_sso_url,
            attribute_mapping: self.attribute_mapping,
            created_at: self.created_at,
            updated_at: self.updated_at,
        }
    }
}

File: services/identity-federation-service/internal/models/sso.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// SsoProvider mirrors `models::sso::SsoProvider`. Persistence-shaped
// row used by the OIDC + SAML SSO flows. Slice 5a wired the OIDC
// surface; slice 5b adds SAML configuration columns
// (`saml_metadata_url`, `saml_entity_id`, `saml_sso_url`,
// `saml_certificate`).
//
// `AttributeMapping` is the JSON column that maps SAML attribute
// names → canonical claim slots (`subject`, `email`, `name`). The
// SAML domain package reads it via `gjson`-style lookups against the
// raw bytes — empty / missing keys fall back to defaults
// (`NameID`, `email`, `name`).
type SsoProvider struct {
	ID                uuid.UUID       `json:"id"`
	Slug              string          `json:"slug"`
	Name              string          `json:"name"`
	ProviderType      string          `json:"provider_type"`
	Enabled           bool            `json:"enabled"`
	ClientID          *string         `json:"client_id,omitempty"`
	ClientSecret      *string         `json:"client_secret,omitempty"`
	IssuerURL         *string         `json:"issuer_url,omitempty"`
	AuthorizationURL  *string         `json:"authorization_url,omitempty"`
	TokenURL          *string         `json:"token_url,omitempty"`
	UserinfoURL       *string         `json:"userinfo_url,omitempty"`
	Scopes            []string        `json:"scopes"`
	SamlMetadataURL   *string         `json:"saml_metadata_url,omitempty"`
	SamlEntityID      *string         `json:"saml_entity_id,omitempty"`
	SamlSsoURL        *string         `json:"saml_sso_url,omitempty"`
	SamlCertificate   *string         `json:"saml_certificate,omitempty"`
	AttributeMapping  json.RawMessage `json:"attribute_mapping,omitempty"`
	CreatedAt         time.Time       `json:"created_at"`
	UpdatedAt         time.Time       `json:"updated_at"`
}

// SsoProviderResponse mirrors `models::sso::SsoProviderResponse` — the
// public-safe view that drops `client_secret` and replaces it with
// `client_secret_configured`.
type SsoProviderResponse struct {
	ID                       uuid.UUID       `json:"id"`
	Slug                     string          `json:"slug"`
	Name                     string          `json:"name"`
	ProviderType             string          `json:"provider_type"`
	Enabled                  bool            `json:"enabled"`
	ClientID                 *string         `json:"client_id,omitempty"`
	ClientSecretConfigured   bool            `json:"client_secret_configured"`
	IssuerURL                *string         `json:"issuer_url,omitempty"`
	AuthorizationURL         *string         `json:"authorization_url,omitempty"`
	TokenURL                 *string         `json:"token_url,omitempty"`
	UserinfoURL              *string         `json:"userinfo_url,omitempty"`
	Scopes                   []string        `json:"scopes"`
	SamlMetadataURL          *string         `json:"saml_metadata_url,omitempty"`
	SamlEntityID             *string         `json:"saml_entity_id,omitempty"`
	SamlSsoURL               *string         `json:"saml_sso_url,omitempty"`
	AttributeMapping         json.RawMessage `json:"attribute_mapping,omitempty"`
	CreatedAt                time.Time       `json:"created_at"`
	UpdatedAt                time.Time       `json:"updated_at"`
}

// IntoResponse mirrors `SsoProvider::into_response` — strips the
// secret while preserving every other field.
func (p *SsoProvider) IntoResponse() SsoProviderResponse {
	return SsoProviderResponse{
		ID:                     p.ID,
		Slug:                   p.Slug,
		Name:                   p.Name,
		ProviderType:           p.ProviderType,
		Enabled:                p.Enabled,
		ClientID:               p.ClientID,
		ClientSecretConfigured: p.ClientSecret != nil && *p.ClientSecret != "",
		IssuerURL:              p.IssuerURL,
		AuthorizationURL:       p.AuthorizationURL,
		TokenURL:               p.TokenURL,
		UserinfoURL:            p.UserinfoURL,
		Scopes:                 p.Scopes,
		SamlMetadataURL:        p.SamlMetadataURL,
		SamlEntityID:           p.SamlEntityID,
		SamlSsoURL:             p.SamlSsoURL,
		AttributeMapping:       p.AttributeMapping,
		CreatedAt:              p.CreatedAt,
		UpdatedAt:              p.UpdatedAt,
	}
}
